### PR TITLE
Builtin Outputs 

### PR DESCRIPTION
**Pull request by nithinmuthukumar**
This PR is aiming to solve #246 by making builtin commands return `CmdOutput`. 
The output of each builtin is captured into a CmdOutput and can be sent to hooks.

### CODE REVIEW

File: crates/shrs_core/src/builtin/debug.rs [hunk 1]
```diff
@@ -28,20 +31,25 @@ impl BuiltinCmd for DebugBuiltin {
...
         args: &Vec<String>,
-    ) -> anyhow::Result<BuiltinStatus> {
+    ) -> anyhow::Result<CmdOutput> {
+        let mut out = String::new();
         let cli = Cli::try_parse_from(args)?;
 
         match &cli.command {
             None => {
-                println!("debug utility");
+                let s = "debug utility\n";
```
Comment (MrPicklePinosaur) [inline crates/shrs_core/src/builtin/debug.rs:40]: using "\n" may not work on all platforms (iirc windows wants "\r" and macos wants "\r\n" for example). this is not very much of an issue for now but maybe make a little note here

File: crates/shrs_core/src/builtin/debug.rs [hunk 2]
```diff
@@ -28,20 +31,25 @@ impl BuiltinCmd for DebugBuiltin {
...
+                print!("{s}");
+                out += s;
             },
             Some(Commands::Env) => {
                 for (var, val) in rt.env.iter() {
-                    println!("{:?} = {:?}", var, val);
+                    let envs = format!("{:?} = {:?}\n", var, val);
+                    print!("{envs}");
+                    out += envs.as_str();
```
Comment (MrPicklePinosaur) [inline crates/shrs_core/src/builtin/debug.rs:48]: i think that its a bit silly having to append to an output string. its also easy for users to use println by accident. i suggest perhaps to make some sort of wrapper class that helps with output. what we want in this case is a stringstream of some sort

maybe your command output writer can implement [BufWriter](https://doc.rust-lang.org/std/io/struct.BufWriter.html) trait. this definitely should use some more investigation

**Review by MrPicklePinosaur — CHANGES_REQUESTED** [review]
looks good, definitely needs more user friendly and less error prone way of outputting though

File: crates/shrs_core/src/builtin/debug.rs (hunk 1 above)
Comment (nithinmuthukumar) [reply crates/shrs_core/src/builtin/debug.rs:40]: Noted, will default to \r\n for now

**Comment by nithinmuthukumar** [comment]
Now this PR is aiming to use `OutputWriter` to simplify the collection of output by making it automatic. Shell.rs automatically calls `ctx.out.begin_collecting()` and `ctx.out.end_collecting()`. All output should now be done using OutputWriter

File: crates/shrs_core/src/builtin/cd.rs [hunk 3]
```diff
@@ -35,8 +36,8 @@ impl BuiltinCmd for CdBuiltin {
                 if let Ok(old_pwd) = rt.env.get("OLDPWD") {
                     PathBuf::from(old_pwd)
                 } else {
-                    eprintln!("no OLDPWD");
-                    return Ok(BuiltinStatus::error());
+                    ctx.out.eprintln("no OLDPWD")?;
```
Comment (MrPicklePinosaur) [inline crates/shrs_core/src/builtin/cd.rs:39]: does this still have the ability to format strings? is it still a variadic function

**Review by MrPicklePinosaur — APPROVED** [review]
looks good. maybe some further thought can be put in on a unified output system in the future but this can do for now

File: crates/shrs_core/src/builtin/cd.rs (hunk 3 above)
Comment (nithinmuthukumar) [reply crates/shrs_core/src/builtin/cd.rs:39]: will make it variadic

**Pull request merged**